docs/porting/target/static_pinmap.md

```diff
@@ -0,0 +1,183 @@
+    pin_function(pinmap->pin, pinmap->function);
+    pin_mode(pinmap->pin, PullNone);
+
+    // Some additional init code
+}
+
+void xxx_init(xxx_t *obj, PinName pin)
+{
+    int peripheral = (int)pinmap_peripheral(pin, PinMap_xxx);
+    int function = (int)pinmap_find_function(pin, PinMap_xxx);
+
+    const PinMap static_pinmap = {pin, peripheral, function};
+
+    xxx_init_direct(obj, &static_pinmap);
+}
+```
+
+- Provide `constexpr` pin-map tables in the header file.
+
+Move pinmap tables from `PeripheralPins.c` to `PeripheralPinMaps.h` (create new file) and add `constexpr` specifier in the pin-map table declarations.
+The tables are required in the header file, so can be included and used by constant expression utility functions to find and return mapping without pulling the pin-map table into the image.
+
+**Note:**
+Please include `<mstd_cstddef>` module and use `MSTD_CONSTEXPR_OBJ_11` macro instead `constexpr` specifier. This must be done for backward compatibility with `ARM 5` compiler which does not support constant expressions. When `ARM 5` compiler is in use `MSTD_CONSTEXPR_OBJ_11` will be translated to `const`.  
+
+Example pin-map table below:
+
+```
+#include <mstd_cstddef>
+
+MSTD_CONSTEXPR_OBJ_11 PinMap PinMap_ADC[] = {
+    {P0_23, ADC0_SE0,    0},
+    {P0_10, ADC0_SE1,    0},
+    {P0_31, ADC0_SE3,    0},
+    {P1_8,  ADC0_SE4,    0},
+    {P2_0,  ADC0_SE5,    0},
+    {P2_13, ADC0_SE6,    0},
+    {P2_11, ADC0_SE7,    0},
+    {NC   , NC      ,    0}
+};
+
+```
+
+- Provide macros for pin-map tables  
+
+Since pin-map table names are not common across all targets the following macros for available pin-map tables are required in `PeripheralPinMaps.h` file:
+
+```
+#define PINMAP_ANALOGIN [PinMap ADC]
+#define PINMAP_ANALOGOUT [PinMap DAC]
+#define PINMAP_I2C_SDA [PinMap I2C SDA]
+#define PINMAP_I2C_SCL [PinMap I2C SCL]
+#define PINMAP_UART_TX [PinMap UART TX]
+#define PINMAP_UART_RX [PinMap UART RX]
+#define PINMAP_UART_CTS [PinMap UART CTS]
+#define PINMAP_UART_RTS [PinMap UART RTS]
+#define PINMAP_SPI_SCLK [PinMap SPI SCLK]
+#define PINMAP_SPI_MOSI [PinMap SPI MOSI]
+#define PINMAP_SPI_MISO [PinMap SPI MISO]
+#define PINMAP_SPI_SSEL [PinMap SPI SSEL]
+#define PINMAP_PWM [PinMap PWM]
+#define PINMAP_QSPI_DATA0 [PinMap QSPI DATA0]
+#define PINMAP_QSPI_DATA1 [PinMap QSPI DATA1]
+#define PINMAP_QSPI_DATA2 [PinMap QSPI DATA2]
+#define PINMAP_QSPI_DATA3 [PinMap QSPI DATA3]
+#define PINMAP_QSPI_SCLK [PinMap QSPI SCLK]
+#define PINMAP_QSPI_SSEL [PinMap QSPI SSEL]
+#define PINMAP_CAN_RD [PinMap CAN RD]
+#define PINMAP_CAN_TD [PinMap CAN RD]
+```
+
+- Provide `STATIC_PINMAP_READY` macro in `PinNames.h`  
+
+Adding this macro will enable the static pin-map support for the target.
+
+```
+/* If this macro is defined, then constexpr utility functions for pin-map seach can be used. */
+#define STATIC_PINMAP_READY 1
+```
+
+### Example usage/testing
+
+Use code below to check if static pinmap extension works.
+
+```
+int main()
```

Comment:
Do we need (or maybe already have) a test for that in HAL suite?

Reply:
I added the test cases which use static pinmap to the FPGA tests. Unfortunately, we do not have tested `CAN`, `QSPI` and `AnalogOut` (this should be added later, because for now `Analogout` test has been removed due to some issue with the FPGA test shield).